edx_event_bus_kafka/publishing/event_producer.py

```diff
@@ -0,0 +1,177 @@
+        event_data: The event data sent to a signal (the kwargs dictionary)
+        event_key_field: Period-delimited string naming the dictionary keys to descend to find the event key data
+
+    Returns:
+        Key data, which might be an integer, string, dictionary, etc.
+    """
+    field_path = event_key_field.split(".")
+    current_data = event_data
+    for field_name in field_path:
+        if isinstance(current_data, dict):
+            if field_name not in current_data:
+                raise Exception(
+                    f"Could not extract key from event; lookup in {event_key_field} "
+                    f"failed at {field_name!r} in dictionary"
+                )
+            current_data = current_data[field_name]
+        else:
+            if not hasattr(current_data, field_name):
+                raise Exception(
+                    f"Could not extract key from event; lookup in {event_key_field} "
+                    f"failed at {field_name!r} in object"
+                )
+            current_data = getattr(current_data, field_name)
+    return current_data
+
+
+def descend_avro_schema(serializer_schema: dict, field_path: List[str]) -> dict:
+    """
+    Extract a subfield within an Avro schema, recursively.
+
+    Arguments:
+        serializer_schema: An Avro schema (nested dictionaries)
+        field_path: List of strings matching the 'name' of successively deeper subfields
+
+    Returns:
+        Schema for some field
+
+    TODO: Move to openedx_events.event_bus.avro.serializer?
+    """
+    subschema = serializer_schema
+    for field_name in field_path:
+        try:
+            # Either descend into .fields (for dictionaries) or .type.fields (for classes).
+            if 'fields' not in subschema:
+                # Descend through .type wrapper first
+                subschema = subschema['type']
+            field_list = subschema['fields']
+
+            matching = [field for field in field_list if field['name'] == field_name]
+            subschema = matching[0]
+        except BaseException as e:
+            raise Exception(
+                f"Error traversing Avro schema along path {field_path!r}; failed at {field_name!r}."
+            ) from e
+    return subschema
+
+
+def extract_key_schema(signal_serializer: AvroSignalSerializer, event_key_field: str) -> str:
+    """
+    From a signal's serializer, extract just the part of the Avro schema that will be used for the Kafka event key.
+
+    Arguments:
+        signal_serializer: The signal serializer to extract a sub-schema from
+        event_key_field: Period-delimited string naming the field 'name' keys to descend to find the key schema
+
+    Returns:
+        The key's schema, as a string.
+    """
+    subschema = descend_avro_schema(signal_serializer.schema, event_key_field.split("."))
+    # Same as used by AvroSignalSerializer#schema_string in openedx-events
+    return json.dumps(subschema, sort_keys=True)
+
+
+@lru_cache
```

Comment:
I need to think through some possible issues with caching vs. config loading. Specifically, if we cache here, then config is not runtime-adjustable with remote config. :-/ Maybe just get rid of this for now.